_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,342 @@
+Using JavaScript, that would look like this:
+
+```js
+fetch("https://api.github.com/events")
+    .then(responseA => {
+        return responseA.json()
+    })
+    .then(events => {
+        if (events.length == 0) {
+            throw "No events."
+        }
+        const { actor : { login } } = events[0]
+        return fetch(`https://api.github.com/users/${login}`)
+    })
+    .then(responseB => {
+        return responseB.json()
+    })
+    .then(user => {
+        if (!user.name) {
+            console.log("unspecified")
+        } else {
+            console.log(user.name)
+        }
+    })
+    .catch(err => {
+        console.error(err)
+    })
+```
+
+It would get a little fancier using `async/await`:
+
+```js
+try {
+    const responseA = await fetch("https://api.github.com/events")
+    const events = await responseA.json()
+    if (events.length == 0) {
+        throw "No events."
+    }
+    const { actor: { login } } = events[0]
+    const responseB = await fetch(`https://api.github.com/users/${login}`)
+    const user = await responseB.json()
+    if (!user.name) {
+        console.log("unspecified")
+    } else {
+        console.log(user.name)
+    }
+} catch (err) {
+    console.error(err)
+}
+```
+
+This is already complicated code to read and understand, and it's tricky to do using Elm as well. Let's see how to achieve the same understanding exactly what we're doing (we've all blindly copied and pasted code in the past, don't deny).
+
+First, let's write the two requests we need; one for fetching the list of events, the second to obtain a given user's details from her `login`:
+
+```haskell
+import Http
+import Json.Decode as Decode
+
+eventsRequest : Http.Request (List String)
+eventsRequest =
+    Http.get "https://api.github.com/events"
+        (Decode.at [ "actor", "login" ] Decode.string
+            |> Decode.list
+        )
+
+nameRequest : String -> Http.Request String
+nameRequest login =
+    Http.get ("https://api.github.com/users/" ++ login)
+        (Decode.at [ "name" ]
+            (Decode.oneOf
+                [ Decode.string
+                , Decode.null "unspecified"
+                ]
+            )
+        )
+```
+
+These two functions return `Http.Request` with the type of data they'll retrieve from the JSON body of their respective responses. `nameRequest` handles the case where Github users don't have entered their full name yet, so the `name` field might be a `null`; as with the JavaScript version, we then default to `"unspecified"`.
+
+That's good but now we need to execute and chain these two requests, the second one depending on the result of the first one, where we retrieve the `actor.login` value of the event object.
+
+Elm is a pure language, meaning you can't have side effects in your functions (an HTTP request is a huge side effect). So your functions must return *something* that represents a given side effect, then another function must handle the result when a response is received.
```

Comment:
As you seem to assume that the reader doesn't know what a side effect is (as you are precising that an HTTP request is a huge side effect), maybe pointing to some document on side effects? Something like [that](http://elmprogramming.com/pure-functions.html#side-effect) maybe.

Reply:
Wow, this is a very complicated definition of a side effect. To me they're functions that alter things outside of their scope, but I may be missing something. I'm probably gonna use that now I think about it.

Reply:
Yes, agree regarding the link. Just explaining what a side effect is in one sentence could be enough.